sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,201 @@
+az_const_span const AZ_HTTP_METHOD_VERB_TRACE = AZ_CONST_STR("TRACE");
+az_const_span const AZ_HTTP_METHOD_VERB_OPTIONS = AZ_CONST_STR("OPTIONS");
+az_const_span const AZ_HTTP_METHOD_VERB_CONNECT = AZ_CONST_STR("CONNECT");
+az_const_span const AZ_HTTP_METHOD_VERB_PATCH = AZ_CONST_STR("PATCH");
+
+AZ_INLINE az_pair * get_headers_start(az_span const buffer, int16_t const max_url_size) {
+  // 8-byte address alignment
+  return (az_pair *)((uintptr_t)(buffer.begin + max_url_size + 7) & ~(uintptr_t)7);
+}
+
+AZ_INLINE uint16_t get_headers_max(az_span const buffer, az_pair * const headers_start) {
+  // We need to compare both pointers as pointers to the same type (uint8_t*)
+  uint8_t * const buffer_end = buffer.begin + buffer.size;
+  uint8_t * const headers_start_uint8ptr = (uint8_t *)headers_start;
+
+  size_t const max_headers = (buffer_end - headers_start_uint8ptr) / sizeof(az_pair);
+
+  // If there's enough space in the buffer, return the max number of elements it can fit (capped to
+  // uint16_t's max)
+  return buffer_end > headers_start_uint8ptr
+      ? (max_headers > ~(uint16_t)0 ? ~(uint16_t)0 : (uint16_t)max_headers)
+      : 0;
+}
+
 az_result az_http_request_builder_init(
     az_http_request_builder * const p_hrb,
     az_span const buffer,
-    int16_t const max_url_size,
+    uint16_t const max_url_size,
     az_const_span const method_verb,
     az_const_span const initial_url) {
-  (void)p_hrb;
-  (void)buffer;
-  (void)max_url_size;
-  (void)method_verb;
-  (void)initial_url;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_span_is_valid(buffer) || !az_const_span_is_valid(method_verb)
+      || !az_const_span_is_valid(initial_url) || max_url_size < initial_url.size) {
+    return AZ_ERROR_ARG;
+  }
+
+  if (buffer.size < max_url_size) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  AZ_RETURN_IF_FAILED(az_span_set(buffer, '\0')); // zero the buffer; we don't have to do this
+
+  az_span uri_buf = { 0, 0 };
+  AZ_RETURN_IF_FAILED(az_span_copy( // copy URL to buffer
+      (az_span){ .begin = buffer.begin, .size = max_url_size },
+      initial_url,
+      &uri_buf));
+
+  az_pair * const headers_start = get_headers_start(buffer, max_url_size);
+  uint16_t const max_headers = get_headers_max(buffer, headers_start);
+
+  *p_hrb = (az_http_request_builder){ .buffer = buffer,
+                                      .method_verb = method_verb,
+                                      .url = uri_buf,
+                                      .body = { .begin = NULL, .size = 0 },
+                                      .max_url_size = max_url_size,
+                                      .max_headers = max_headers,
+                                      .retry_headers_start = max_headers,
+                                      .headers_end = 0 };
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_set_query_parameter(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
```

Comment:
The naming here should match az_pair. If az_pair uses key/value, then we should use key value too. Maybe az_pair should use name/value though and then can stay as is.

Reply:
Done.